### PR TITLE
feat: impl pubsub in metasrv

### PR DESCRIPTION
**Pull request by Fengys123**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This pr mainly adds the `pub/sub` framework.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/meta-srv/src/pubsub/subscribe_manager.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+    fn un_subscribe(&self, req: UnSubRequest) -> Result<()> {
+        let UnSubRequest { subscriber_id } = req;
+
+        info!("Add a un_subscription, subscriber_id: {}", subscriber_id);
+
+        for entry in self.topic2sub.iter_mut() {
+            let mut sub_list = entry;
+            sub_list.retain(|subscriber| subscriber.id() != subscriber_id)
+        }
```
Comment (MichaelScofield) [inline src/meta-srv/src/pubsub/subscribe_manager.rs:106]: Can it simply:

```suggestion
        for mut entry in self.topic2sub.iter_mut() {
            entry.retain(|subscriber| subscriber.id() != subscriber_id)
        }
```

Comment (Fengys123) [reply src/meta-srv/src/pubsub/subscribe_manager.rs:106]: Good suggestion!

**Comment by Fengys123** [comment]
@MichaelScofield PTAL

File: src/meta-srv/src/metasrv.rs [hunk 2]
```diff
@@ -339,6 +358,8 @@ impl MetaSrv {
         let mailbox = self.mailbox.clone();
         let election = self.election.clone();
         let skip_all = Arc::new(AtomicBool::new(false));
+        let publish = self.publish.clone();
```
Comment (fengjiachun) [inline src/meta-srv/src/metasrv.rs:361]: Suggest placing the publish object into the reporterHandler (when constructing the reporterHandler) for now, and decide whether to put it into ctx in the future according to usage.

File: src/meta-srv/src/metasrv/builder.rs [hunk 3]
```diff
@@ -237,6 +244,8 @@ impl MetaSrvBuilder {
             ddl_manager,
             table_metadata_manager,
             greptimedb_telemerty_task: get_greptimedb_telemetry_task(meta_peer_client).await,
+            publish: Some(publish),
+            subscribe_manager: Some(sub_manager),
```
Comment (fengjiachun) [inline src/meta-srv/src/metasrv/builder.rs:248]: If the two components `publish` and `subscribe_manager` must appear in pairs, that is, either all are `Some` or all are `None`, then I suggest wrapping the two components together.

File: src/meta-srv/src/pubsub.rs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+mod publish;
+mod subscribe_manager;
+mod subscriber;
+#[cfg(test)]
+mod tests;
+
+pub use publish::*;
+pub use subscribe_manager::*;
+pub use subscriber::*;
```
Comment (fengjiachun) [inline src/meta-srv/src/pubsub.rs:25]: It is not recommended to use xxx.*

Comment (Fengys123) [reply src/meta-srv/src/pubsub.rs:25]: why? What is recommended?

Comment (Fengys123) [reply src/meta-srv/src/pubsub.rs:25]: Oh, I got

File: src/meta-srv/src/metasrv.rs (hunk 2 above)
Comment (Fengys123) [reply src/meta-srv/src/metasrv.rs:361]: Good suggestion!

**Comment by Fengys123** [comment]
@fengjiachun PTAL

File: src/servers/Cargo.toml [hunk 5]
```diff
@@ -73,7 +73,7 @@ schemars = "0.8"
 secrecy = { version = "0.8", features = ["serde", "alloc"] }
 serde.workspace = true
 serde_json = "1.0"
-session = { path = "../session" }
+session = { path = "../session", features = ["testing"] }
```
[the remark this answers is not in the document]
Comment (Fengys123) [reply src/servers/Cargo.toml:76]: emmm, let me try.

File: src/meta-srv/src/pubsub/publish.rs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+            .subscribe_manager
+            .subscribers_by_topic(&message.topic());
+
+        for sub in sub_list {
+            if sub.transport_msg(message.clone()).await.is_err() {
+                // If an error occurs, we consider the subscriber offline,
+                // so un_subscribe here.
+                let req = UnSubRequest {
+                    subscriber_id: sub.id(),
+                };
```
[the remark this answers is not in the document]
Comment (Fengys123) [reply src/meta-srv/src/pubsub/publish.rs:64]: The `UnSubRequest` operation disconnect the grpc connection.

File: src/servers/Cargo.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (Fengys123) [reply src/servers/Cargo.toml]: we can add `session = { path = "../session", features = ["testing"] }` in `[dev-dependencies]`

**Review by MichaelScofield — APPROVED** [review]

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**